### PR TITLE
Ported recent profinda changes to the Operation module [PF-6472]

### PR DESCRIPTION
**Pull request by k-rudy**
JIRA: https://profinda.atlassian.net/browse/PF-6472

## What

1. Ported recent changes from ProFinda::Operation
2. Added information copying over for `operation` executor
3. Added minimal dry version requirement

### CODE REVIEW

File: lib/opera/operation/instructions/executors/validate.rb
Comment (petergebala) [inline lib/opera/operation/instructions/executors/validate.rb:18]: we don't need that. I fixed that by adding `.to_h` on the left.

File: lib/opera/operation/instructions/executors/operation.rb
Comment (petergebala) [inline lib/opera/operation/instructions/executors/operation.rb:19]: also, information by design wasn't meant to hold exceptions

Comment (petergebala) [reply lib/opera/operation/instructions/executors/operation.rb:19]: we have errors for that and exceptions arrays

File: lib/opera/operation/instructions/executors/validate.rb
Comment (petergebala) [reply lib/opera/operation/instructions/executors/validate.rb:18]: but please, double-check. At least it works just fine in HAL

File: lib/opera/operation/instructions/executors/operation.rb
Comment (petergebala) [reply lib/opera/operation/instructions/executors/operation.rb:19]: https://github.com/Profinda/opera/commit/e5e83ca1b3fce96e4e85e2d100327f038114d131#diff-828218ef941ed5d36e1ab826614301e149b77cbc896665aaae4b327d3bee4449R17

File: lib/opera/operation/instructions/executors/validate.rb
Comment (petergebala) [reply lib/opera/operation/instructions/executors/validate.rb:18]: https://github.com/Profinda/opera/commit/e5e83ca1b3fce96e4e85e2d100327f038114d131#diff-828218ef941ed5d36e1ab826614301e149b77cbc896665aaae4b327d3bee4449R17

File: lib/opera/operation/instructions/executors/operation.rb
Comment (k-rudy) [reply lib/opera/operation/instructions/executors/operation.rb:19]: It's not for storing exceptions or errors, that's right, but we should keep the information that is collected on the nested operations as that's useful. I'll add it to the success path as well.

Comment (petergebala) [reply lib/opera/operation/instructions/executors/operation.rb]: It might be misleading and not readable. We are adding information stright to parent operations. Imagine situation that you have 5 level deep structure and the same information on every level: `Executed in 5s`. How can you know from which operations it came? If we want to do so (I am not sure about that), we should do a similar thing that we did for `execution`

Comment (k-rudy) [reply lib/opera/operation/instructions/executors/operation.rb]: not sure I agree with this. With current approach we have flexibility to either add to parent or add to child as 
```
add_information(child: { status: 404 })
```

or 

```
add_information(status: 404)
```

In many cases you don't want to case about the actual source of 404 state, but you want to add it.

File: lib/opera/operation/instructions/executors/validate.rb
Comment (k-rudy) [reply lib/opera/operation/instructions/executors/validate.rb]: doesn't work, unfortunately - https://app.circleci.com/pipelines/github/Profinda/profinda_saas/3934/workflows/dce9fe51-3a10-474f-a72a-c6f0fe05d457/jobs/53381

Comment (petergebala) [reply lib/opera/operation/instructions/executors/validate.rb:18]: but the error is about `information`

Comment (petergebala) [reply lib/opera/operation/instructions/executors/validate.rb:18]: no @k-rudy. It works. Problem is that our code is a hack. In step `validate` we expect dry-validation result, not sth else. Here we have custom validator which tries to mimic our intefrace and it breaks. Please update our validator:
https://github.com/Profinda/profinda_saas/blob/master/engines/pro_finda/lib/pro_finda/validator.rb

add line:
```
delegate to_h, to: :errors
```

**Review by petergebala — APPROVED** [review]

**Review by kikorb — APPROVED** [review]

**Comment by kikorb** [comment]
@k-rudy 
Please manually bump the version and tag the master branch with the version once merged and please build and publish that new version to rubygems

**Comment by kikorb** [comment]
Screenshot 2021-03-04 at 12 51 55

Maybe it is worth to setup that GH Action to automate it, @k-rudy @petergebala

**Comment by k-rudy** [comment]
I like the idea, @kikorb but I suggest we do it separately to not block this PR. If you agree, I will raise a story to `Tech debt cleanup` epic

**Review by petergebala — APPROVED** [review]

**Comment by petergebala** [comment]
:shipit:

**Pull request merged**
